File: ruby/lesson 7/binary_search_tree.rb
# https://leetcode.com/problems/search-in-a-binary-search-tree/
# https://leetcode.com/problems/insert-into-a-binary-search-tree/
# https://leetcode.com/problems/delete-node-in-a-bst/

class BinarySearchTree
    def initialize
        @root = nil
    end
    
    def get(val)
        get_value(root, val)
    end
    
    def put(val)
        @root = put_value(root, val)
    end
    
    def delete(val)
        @root = delete_value(root, val)
    end

    private
    
    def get_value(node, val)
        return nil if node.nil?
        
        return get_value(node.left, val) if val < node.val
        return get_value(node.right, val) if val > node.val
        node
    end

    def put_value(node, val)
        return TreeNode.new(val) if node.nil?
        
        node.left = put_value(node.left, val) if val < node.val
        node.right = put_value(node.right, val) if val > node.val

        node
    end

    def delete_value(node, val)
        return nil if node.nil?
        
        if val < node.val
            node.left = delete_value(node.left, val)
        elsif val > node.val
            node.right = delete_value(node.right, val)
        else
            return node.right if node.left.nil?
            return node.left if node.right.nil?
            
            temp = node
            node = min(temp.right)
            node.right = delete_min(temp.right)
            node.left = temp.left
        end
        node
    end
    
    def min(node)
        node.left.nil? ? node : min(node.left)
    end
    
    def delete_min(node)
        return node&.right if node&.left.nil?
        node.left = delete_min(node.left)
        node
    end
end

File: javascript/lesson 7/binarySearchTree.js
// https://leetcode.com/problems/search-in-a-binary-search-tree/
// https://leetcode.com/problems/insert-into-a-binary-search-tree/
// https://leetcode.com/problems/delete-node-in-a-bst/

class BinarySearchTree {
    constructor() {
        this.root = null;
    }
    
    get(val) {
        this.get(this.root, val);
    }
    
    get(node, val) {
        if (!node) return null;
        
        if (val < node.val) return this.get(node.left, val);
        if (val > node.val) return this.get(node.right, val);
        return node;
    }
    
    put(val) {
        this.root = this.put(this.root, val);
    }
    
    put(node, val) {
        if (!node) return new TreeNode(val);

        if (val < node.val)
            node.left = this.put(node.left, val);
        else if (val > node.val)
            node.right = this.put(node.right, val);
        else
            node.val = val;

        return node;
    }   
    
    delete(val) {
        this.root = this.delete(this.root, val);
    }
    
    delete(node, val) {
        if (!node) return null;
        
        if (val < node.val) node.left = this.delete(node.left, val);
        else if (val > node.val) node.right = this.delete(node.right, val);
        else {
            if (node.right == null) return node.left;
            else if (node.left == null) return node.right;
            else {
                const temp = node;
                node = this.min(temp.right);
                node.right = this.deleteMin(temp.right);
                node.left = temp.left;
            }
        }
        return node;
    }
    
    min(node) {
        return !node.left ? node : this.min(node.left); 
    }
    
    deleteMin(node) {
        if (!node.left) return node.right;
        node.left = this.deleteMin(node.left);
        return node;
    }
}
